app/models/manageiq/providers/hawkular/inventory/server_operations.rb

```diff
@@ -0,0 +1,278 @@
+            end
+
+            def domain_operation(name, *args)
+              generic_operation("#{name}_middleware_domain_server", *args)
+            end
+
+            def standalone_operation(name, *args)
+              generic_operation("#{name}_middleware_server", *args)
+            end
+
+            def generic_operation(name, action_name, default_params = {}, default_extra_data = {})
+              define_method(name) do |ems_ref, params = {}, extra_data = {}|
+                run_generic_operation(action_name.to_sym, ems_ref, default_params.merge(params), default_extra_data.merge(extra_data))
+              end
+            end
+
+            def specific_operation(name, action_name, default_params = {})
+              define_method(name) do |ref, params = {}|
+                params[:resourcePath] = ref.to_s
+                run_operation(default_params.merge(params), action_name)
+              end
+            end
+          end
+
+          NotificationArgs = Struct.new(:type, :operation_name, :operation_args, :target_resource, :entity_klass, :detailed_message) do
+            def self.success(*args)
+              new(:mw_op_success, *args)
+            end
+
+            def event_type(entity)
+              attributes = {
+                :entity_type => entity.kind_of?(MiddlewareServer) ? 'MwServer' : 'MwDomain',
+                :operation   => operation_name,
+                :status      => type == :mw_op_success ? 'Success' : 'Failed'
+              }
+
+              '%{entity_type}.%{operation}.%{status}' % attributes
+            end
+
+            def event_message(entity)
+              attributes = {
+                :operation => operation_name,
+                :server    => entity.name,
+                :status    => type == :mw_op_success ? _('succeeded') : _('failed')
+              }
+
+              message = _('%{operation} operation for %{server} %{status}') % attributes
+
+              message + ": #{detailed_message}" if detailed_message
+            end
+          end
+
+          def add_middleware_datasource(ems_ref, hash)
+            with_provider_connection do |connection|
+              datasource_data = {
+                :resourcePath         => ems_ref.to_s,
+                :datasourceName       => hash[:datasource]["datasourceName"],
+                :xaDatasource         => hash[:datasource]["xaDatasource"],
+                :jndiName             => hash[:datasource]["jndiName"],
+                :driverName           => hash[:datasource]["driverName"],
+                :driverClass          => hash[:datasource]["driverClass"],
+                :connectionUrl        => hash[:datasource]["connectionUrl"],
+                :userName             => hash[:datasource]["userName"],
+                :password             => hash[:datasource]["password"],
+                :xaDataSourceClass    => hash[:datasource]["driverClass"],
```

Comment:
the driverClass is used both here and above in line 71 - is this a bug?

Reply:
If it is then it is wortking as such, and outside of the scope of this PR, given that I'm focusing on only refactoring and not changing anything else.